jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/json/DockerConfigTemplate.java

```diff
@@ -97,48 +96,59 @@
    */
   @Nullable
   public String getAuthFor(String registry) {
-    Predicate<String> exactMatch = registry::equals;
-    Predicate<String> withHttps = ("https://" + registry)::equals;
-    Predicate<String> startsWith = name -> name.startsWith(registry);
-    Predicate<String> startsWithAndWithHttps = name -> name.startsWith("https://" + registry);
-    AuthTemplate authTemplate =
-        getAuthTemplate(Arrays.asList(exactMatch, withHttps, startsWith, startsWithAndWithHttps));
-
+    AuthTemplate authTemplate = findFirstInMapByKey(auths, getRegistryMatchersFor(registry));
     return authTemplate != null ? authTemplate.auth : null;
   }
 
-  /** Returns the first {@link AuthTemplate} matching the given predicates (short-circuiting). */
-  private AuthTemplate getAuthTemplate(List<Predicate<String>> registryMatches) {
-    return registryMatches
-        .stream()
-        .map(this::getAuthTemplate)
-        .filter(Objects::nonNull)
-        .findFirst()
-        .orElse(null);
-  }
-
-  /** Returns {@link AuthTemplate} matching the given predicate. */
-  private AuthTemplate getAuthTemplate(Predicate<String> registryMatch) {
-    return auths.keySet().stream().filter(registryMatch).map(auths::get).findFirst().orElse(null);
-  }
-
   /**
+   * Returns {@code credsStore} or {@code credHelpers} for the given {@code registry}. If there
+   * exists a matching registry entry in {@code auths}, returns {@code credStore}; otherwise, a
+   * matching entry in {@code credHelpers} is returned based on the following lookup order:
+   *
+   * <ol>
+   *   <li>Exact registry name
+   *   <li>https:// + registry name
+   *   <li>registry name + arbitrary suffix
+   *   <li>https:// + registry name + arbitrary suffix
+   * </ol>
+   *
    * @param registry the registry to get the credential helpers for
    * @return {@code credsStore} if {@code registry} is present in {@code auths}; otherwise, searches
    *     {@code credHelpers}; otherwise, {@code null} if not found
    */
   @Nullable
   public String getCredentialHelperFor(String registry) {
     if (credsStore != null) {
-      // The registry could be prefixed with the HTTPS protocol.
-      if (auths.containsKey(registry) || auths.containsKey("https://" + registry)) {
+      // The registry could be prefixed with the HTTPS protocol and/or have suffixes.
```

Comment:
Should we also handle aliases here?

Reply:
I don't think so. Aliases would be handled at a higher-level, from callers.

If you are talking about suffixes, `startsWith()` will cover both the exact and suffix matches. However, I just found a pitfall (https://github.com/GoogleContainerTools/jib/issues/586#issuecomment-405960261), so I think I'll just reuse `getRegistryMatchersFor()`, which actually simplifies the code.